File: cpp/sudoku_solver.cpp
//Backtracking approach to solve Sudoku
#include <vector>
#include <iostream>

using namespace std;

bool isSafe(const vector<vector<char>>& board, int row, int col, char num) {
    for (int i = 0; i < 9; ++i) {
        if (board[row][i] == num || board[i][col] == num) {
            return false;
        }
    }

    // Check 3x3 box
    int startRow = row - row % 3;
    int startCol = col - col % 3;
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            if (board[i + startRow][j + startCol] == num) {
                return false;
            }
        }
    }
    return true;
}

bool solveSudoku(vector<vector<char>>& board) {
    for (int i = 0; i < 9; ++i) {
        for (int j = 0; j < 9; ++j) {
            if (board[i][j] == '.') {
                
                // Try digits from 1-9
                for (char num = '1'; num <= '9'; ++num) {
                    if (isSafe(board, i, j, num)) {
                        board[i][j] = num; // Place the number

                        if (solveSudoku(board)) {
                            return true; // Solved.
                        }

                        board[i][j] = '.'; // Backtrack takes place. Tries next number.
                    }
                }
                return false; // No number worked in this cell
            }
        }
    }
    return true; // No empty cells left, puzzle is solved
}
// Time Complexity: Exponential in the worst case, but efficient for solvable Sudoku problems.

File: SudokuSolver/sudoku_solver.py
"""
Sudoku Solver
Author: Tathagat Gaikwad
Contribution for Hacktoberfest 2025
Repository: https://github.com/strivedi4u/hacktoberfest2025
"""

def print_board(board):
    """Prints the Sudoku board in a readable 9x9 grid format."""
    for i in range(len(board)):
        if i % 3 == 0 and i != 0:
            print("- - - - - - - - - - - -")
        for j in range(len(board[0])):
            if j % 3 == 0 and j != 0:
                print(" | ", end="")
            if j == 8:
                print(board[i][j])
            else:
                print(str(board[i][j]) + " ", end="")


def find_empty(board):
    """Finds the next empty cell in the Sudoku board (denoted by 0)."""
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] == 0:
                return (i, j)  # row, col
    return None


def valid(board, num, pos):
    """Checks whether a number placement is valid."""
    row, col = pos

    # Check row
    if num in board[row]:
        return False

    # Check column
    for i in range(len(board)):
        if board[i][col] == num:
            return False

    # Check 3x3 box
    box_x = col // 3
    box_y = row // 3

    for i in range(box_y * 3, box_y * 3 + 3):
        for j in range(box_x * 3, box_x * 3 + 3):
            if board[i][j] == num:
                return False

    return True


def solve(board):
    """Solves the Sudoku board using backtracking."""
    find = find_empty(board)
    if not find:
        return True  # Solved
    else:
        row, col = find

    for num in range(1, 10):
        if valid(board, num, (row, col)):
            board[row][col] = num

            if solve(board):
                return True

            board[row][col] = 0  # backtrack

    return False


if __name__ == "__main__":
    # Example Sudoku board (0 denotes empty cells)
    sudoku_board = [
        [7, 8, 0, 4, 0, 0, 1, 2, 0],
        [6, 0, 0, 0, 7, 5, 0, 0, 9],
        [0, 0, 0, 6, 0, 1, 0, 7, 8],
        [0, 0, 7, 0, 4, 0, 2, 6, 0],
        [0, 0, 1, 0, 5, 0, 9, 3, 0],
        [9, 0, 4, 0, 6, 0, 0, 0, 5],
        [0, 7, 0, 3, 0, 0, 0, 1, 2],
        [1, 2, 0, 0, 0, 7, 4, 0, 0],
        [0, 4, 9, 2, 0, 6, 0, 0, 7]
    ]

    print("Sudoku Board (Before Solving):")
    print_board(sudoku_board)
    solve(sudoku_board)
    print("\nSudoku Board (After Solving):")
    print_board(sudoku_board)
